File: 30.串联所有单词的子串.java
/*
 * @lc app=leetcode.cn id=30 lang=java
 *
 * [30] 串联所有单词的子串
 */

// @lc code=start
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> rs = new ArrayList<>();
        if(words.length == 0){
            return rs;
        }
        int one_word = words[0].length();
        int allLen = one_word * words.length;
        Map<String, Integer> map = new HashMap<>();
        for(String w : words){
            map.put(w, map.getOrDefault(w, 0) + 1);
        }
        int i = 0;
        while(i + allLen <= s.length()){
            Map<String, Integer> temp = new HashMap<>();
            for(int j = 0; j < words.length; j++){
                String w = s.substring(i + j * one_word, i + (j + 1) * one_word);
                temp.put(w, temp.getOrDefault(w, 0) + 1);
            }
            if(temp.equals(map)){
                rs.add(i);
            }
            i++;
        }
        return rs;
    }
}
// @lc code=end

File: 30.串联所有单词的子串.go
/*
 * @lc app=leetcode.cn id=30 lang=golang
 *
 * [30] 串联所有单词的子串
 */

// @lc code=start
func findSubstring(s string, words []string) []int {
	res := make([]int, 0)
	slen, wordLen := len(s), len(words)

	if slen == 0 || wordLen == 0 {
		return res
	}

	singleLen := len(words[0])

	m := map[string]int{}
	for i := 0; i < wordLen; i++ {
		m[words[i]]++
	}

	for i := 0; i < slen-singleLen*wordLen+1; i++ {
		tempMap := map[string]int{}
		for l := 0; l < wordLen; l++ {
			subS := s[i+l*singleLen : i+(l+1)*singleLen]
			if _, ok := m[subS]; !ok {
				break
			}
			tempMap[subS]++
			if tempMap[subS] > m[subS] {
				break
			}
		}
		if checkMap(m, tempMap) {
			res = append(res, i)
		}
	}
	return res
}

func checkMap(x, y map[string]int) bool {
	for k, v := range x {
		if yV, ok := y[k]; ok {
			if v != yV {
				return false
			}
		} else {
			return false
		}
	}
	return true
}

// @lc code=end
